### PR TITLE
[DistillationTrainer refactor] Remove `lmbda` and the off-policy training branch

### PR DESCRIPTION
**Pull request by qgallouedec**
*Plan step 2.3 — stacked on PR #6459. Completes Phase 2. Part of #6449.*

The trainer is now **always on-policy**: every generation-batch slice is generated by the student; there is no dataset-completion (off-policy) path.

### Trainer
- `_fill_buffer` generates for every slice. The per-slice on/off-policy coin flip (`random.random() <= lmbda`, broadcast across ranks) and the "store the dataset slice directly" off-policy branch are gone.
- Remove `self.lmbda`, `self._buffered_on_policy_flags`, and the four on/off-policy loss accumulators. `training_step` drops the on/off-policy attribution (its completion-length metrics lose the `on_policy_` / `off_policy_` prefix); `log` drops the `on_policy_loss` / `off_policy_loss` emission and the accumulator all-reduce.
- Drop the now-unused `broadcast_object_list`, `torch.distributed as dist`, and `DistributedType` imports (`random` stays — rich sampling uses it).

### Config
- Remove `lmbda`, its range check, the deprecation warning from #6458, and the `num_generations > 1 and lmbda < 1.0` warning. `warnings` import removed (now unused).

### Tests
- `_make_args` drops `lmbda=0.0`, so every training test now runs on-policy (the student generates). The on/off-policy test collapses to a single `test_train_updates_params`. The two `lmbda` deprecation tests are removed. The server ragged-grad test drops `lmbda=0.0`.

### Verification (pytest)

**49 passed, 1 xfailed.** Two failures surfaced first and were fixed in this PR: `_make_args` and the server test both passed `lmbda=…` to a config that no longer accepts it. Warnings dropped 10 → 1 (the deprecation is gone with the feature).

### Behaviour change

`_make_args` previously defaulted to `lmbda=0.0` (off-policy, dataset completions, no generation). Tests now generate on-policy on CPU, so the suite is somewhat slower. Metric names `completions/{on,off}_policy_*` become `completions/*` — worth noting for anyone watching those keys; they realign with GRPO's naming in Phase 5.

### Risk

Medium. Removes a whole training mode and rewrites `_fill_buffer` / `training_step` / `log`. Guarded by the always-on-policy training test and the full green suite. The off-policy mode lives on in `GKDTrainer` (per #6459's paper-index redirect).

---

> [!NOTE]
> **Medium Risk**
> Removes an entire training mode and rewrites buffering, training_step, and logging; default test behavior shifts from off-policy to on-policy (slower tests, metric key changes for dashboards).
> 
> **Overview**
> **DistillationTrainer** is now **always on-policy**: every gradient-accumulation slice gets student-generated completions; the off-policy path that trained on dataset assistant completions is removed.
> 
> **Config:** `lmbda` and its validation, deprecation warning, and `num_generations` + off-policy warning are dropped from `DistillationConfig`.
> 
> **Trainer:** `_fill_buffer` always calls generation for all slices (no per-slice random mix or `broadcast_object_list`). Removed on/off-policy loss accumulators, `on_policy_loss` / `off_policy_loss` logs, and distributed all-reduce for those. Completion metrics rename from `completions/{on,off}_policy_*` to `completions/*`. Unused `dist` / `DistributedType` imports removed.
> 
> **Tests:** `lmbda` deprecation tests removed; on/off-policy parametrized test collapsed to `test_train_updates_params`; test helpers no longer pass `lmbda` (defaults were off-policy before, so the suite now exercises generation on CPU).
> 
> Off-policy mixing remains available on **`GKDTrainer`**.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit ddcb836a4d885cfcbee190f583eb55a190d06320. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: trl/experimental/distillation/distillation_trainer.py [hunk 1]
```diff
@@ -300,8 +299,8 @@ class DistillationTrainer(_BaseTrainer):
 
     Supports:
     - Generalized JSD loss (forward KL, reverse KL, or interpolated JSD via `beta`)
-    - On-policy / off-policy mixing via `lmbda` (buffered across gradient accumulation)
-    - Local teacher model or external teacher via vLLM server
+    - On-policy distillation: the student generates completions, the teacher scores them
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/experimental/distillation/distillation_trainer.py:302]: The quick-start and example script are updated together with the rest of the user-facing docs in one dedicated pass at promotion (#6449, items 56–58), rather than per-PR. Deferring.

File: trl/experimental/distillation/distillation_trainer.py [hunk 2]
```diff
@@ -697,43 +690,26 @@ def _prepare_inputs(self, generation_batch: dict[str, torch.Tensor | Any]) -> di
 
     @profiling_decorator
     def _fill_buffer(self, generation_batch: dict[str, torch.Tensor | Any], buffer_steps: int):
-        """Split batch into slices and decide which are on-policy (student-generated) vs off-policy."""
+        """Split the batch into slices and generate student completions for each (always on-policy)."""
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/experimental/distillation/distillation_trainer.py:691]: https://github.com/huggingface/trl/pull/6460/commits/26038ddee73f4c1b7b47e1043a20f56bd18b7bb8

**Review by cmpatino — APPROVED** [review]
The only issue I'm finding, as with previous PRs, is that the docs in distillation_trainer.md become outdated.

In this one, we still pass the `lmbda` param in the example.

Maybe it's easier if I skip the comments about the docs and defer them to the last stage of the refactor. WDYT?

File: tests/experimental/test_distillation_trainer.py [hunk 3]
```diff
@@ -266,17 +253,11 @@ def test_distillation_trainer_train_runs_with_local_teacher(self):
...
-
-        `lmbda=0.0` trains on the dataset's own completions, `lmbda=1.0` on completions the student generates. The
-        trainer is scheduled to become always-on-policy, so the off-policy case is pinned here to make its removal a
-        deliberate deletion rather than a silent one.
-        """
+    def test_train_updates_params(self):
+        """Training is always on-policy: the student generates completions, the teacher scores them, params move."""
         # Higher lr than the default: gradients are tiny on this model and the default lr can stall the update, which
         # would make the assertion below vacuous.
-        trainer = self._make_local_trainer(lmbda=lmbda, max_steps=2, learning_rate=0.1)
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply tests/experimental/test_distillation_trainer.py:279]: https://github.com/huggingface/trl/pull/6460/commits/ddcb836a4d885cfcbee190f583eb55a190d06320
